src/openrct2/scripting/ScriptEngine.h

```diff
@@ -0,0 +1,213 @@
+/*****************************************************************************
+ * Copyright (c) 2014-2018 OpenRCT2 developers
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/OpenRCT2/OpenRCT2
+ *
+ * OpenRCT2 is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+#ifdef __ENABLE_SCRIPTING__
+
+#    include "../common.h"
+#    include "../core/FileWatcher.h"
+#    include "../world/Location.hpp"
+#    include "HookEngine.h"
+#    include "Plugin.h"
+
+#    include <future>
+#    include <memory>
+#    include <mutex>
+#    include <queue>
+#    include <string>
+#    include <unordered_map>
+#    include <unordered_set>
+#    include <vector>
+
+struct duk_hthread;
+typedef struct duk_hthread duk_context;
+
+struct GameAction;
+class GameActionResult;
+class FileWatcher;
+class InteractiveConsole;
+
+namespace OpenRCT2
+{
+    interface IPlatformEnvironment;
+}
+
+namespace OpenRCT2::Scripting
+{
+    class ScriptExecutionInfo
+    {
+    private:
+        std::shared_ptr<Plugin> _plugin;
+        bool _isGameStateMutable{};
+
+    public:
+        class PluginScope
+        {
+        private:
+            ScriptExecutionInfo& _execInfo;
+            std::shared_ptr<Plugin> _plugin;
+
+        public:
+            PluginScope(ScriptExecutionInfo& execInfo, std::shared_ptr<Plugin> plugin, bool isGameStateMutable)
+                : _execInfo(execInfo)
+                , _plugin(plugin)
+            {
+                _execInfo._plugin = plugin;
+                _execInfo._isGameStateMutable = isGameStateMutable;
+            }
+            PluginScope(const PluginScope&) = delete;
+            ~PluginScope()
+            {
+                _execInfo._plugin = nullptr;
+                _execInfo._isGameStateMutable = false;
+            }
+        };
+
+        std::shared_ptr<Plugin> GetCurrentPlugin()
+        {
+            return _plugin;
+        }
+
+        bool IsGameStateMutable()
+        {
+            return _isGameStateMutable;
+        }
+    };
+
+    class DukContext
+    {
+    private:
+        duk_context* _context{};
+
+    public:
+        DukContext();
+        DukContext(DukContext&) = delete;
+        DukContext(DukContext&& src) noexcept
+            : _context(std::move(src._context))
+        {
+        }
```

Comment:
This won't zero the `src._context` pointer. I'm not sure that is what you want, is it?
```suggestion
        DukContext(DukContext&& src) noexcept
            : _context(std::move(src._context))
        {
            src._context = {};
        }
```

Reply:
This sounds like a real source of bugs, are you fixing it Ted?